File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        diameter=0

        def maxDiameter(node):
            nonlocal diameter
            if not node:
                return 0
            left = maxDiameter(node.left)
            right = maxDiameter(node.right)
            diameter = max(diameter,left+right)
            return max(left,right)+1
        
        maxDiameter(root)
        return diameter
        

File: Data Structures & Algorithms/max-water-container/submission-0.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
      int max_area=INT_MIN;
      int left=0,right=heights.size()-1;
      while(left<right){
        max_area=max(min(heights[left],heights[right])*(right-left),max_area);
        if(heights[left]<heights[right]){
            left++;
        }
        else{
            right--;
        }
        
      }  
      return max_area;
    }
};
